File: app/src/androidTest/java/ru/iteco/fmhandroid/ui/AuthTest.java
package ru.iteco.fmhandroid.ui;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.RootMatchers.withDecorView;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.not;

import android.view.View;

import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import io.qameta.allure.Description;
import io.qameta.allure.Epic;
import io.qameta.allure.Feature;
import io.qameta.allure.Severity;
import io.qameta.allure.SeverityLevel;
import io.qameta.allure.Story;
import io.qameta.allure.android.rules.ScreenshotRule;
import ru.iteco.fmhandroid.R;
import ru.iteco.fmhandroid.data.TestData;
import ru.iteco.fmhandroid.page.LoginPage;
import ru.iteco.fmhandroid.page.MainPage;

@Epic("User Management")
@Feature("Authentication")
@LargeTest
@RunWith(AndroidJUnit4.class)
public class AuthTest {

    @Rule
    public ScreenshotRule screenshotRule = new ScreenshotRule(ScreenshotRule.Mode.FAILURE, "ss_on_failure");
    @Rule
    public ActivityScenarioRule<AppActivity> activityRule =
            new ActivityScenarioRule<>(AppActivity.class);

    @Before
    public void logoutIfNeeded() {
        if (mainPage.isMainScreenDisplayed()) {
            mainPage.logout();
        }
    }

    @Before
    public void setUp() {
        activityRule.getScenario().onActivity(activity ->
                decorView = activity.getWindow().getDecorView()
        );
    }

    private final LoginPage loginPage = new LoginPage();
    private final MainPage mainPage = new MainPage();
    private View decorView;

    @Test
    @Story("Login with valid credentials")
    @Severity(SeverityLevel.BLOCKER)
    @Description("Verify that a user can successfully log in with valid login and password")
    public void tc001_validLogin() {
        loginPage
                .waitForLoginScreen()
                .enterLogin(TestData.VALID_LOGIN)
                .enterPassword(TestData.VALID_PASSWORD)
                .tapLoginButton();
    }

    @Test
    @Story("Login with empty fields")
    @Severity(SeverityLevel.NORMAL)
    @Description("Verify the error message when login and password fields are left blank")
    public void tc002_emptyLoginAndPassword() {
        loginPage
                .waitForLoginScreen()
                .enterLogin("")
                .enterPassword("")
                .tapLoginButtonExpectingError();

        onView(withText(R.string.empty_login_or_password))
                .inRoot(withDecorView(not(is(decorView))))
                .check(matches(isDisplayed()));
    }

    @Test
    @Story("Login with invalid credentials")
    @Severity(SeverityLevel.CRITICAL)
    @Description("Verify the error message when entering non-existent user data")
    public void tc003_invalidLogin() {
        loginPage
                .waitForLoginScreen()
                .enterLogin(TestData.INVALID_LOGIN)
                .enterPassword(TestData.VALID_PASSWORD)
                .tapLoginButtonExpectingError();

        onView(withText(R.string.error))
                .inRoot(withDecorView(not(is(decorView))))
                .check(matches(isDisplayed()));
    }
}

File: app/src/androidTest/java/ru/iteco/fmhandroid/ui/AuthTest.kt
package ru.iteco.fmhandroid.ui

import android.view.View
import androidx.test.espresso.Espresso.closeSoftKeyboard
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.NoMatchingViewException
import androidx.test.espresso.action.ViewActions.replaceText
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.RootMatchers.withDecorView
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import org.hamcrest.Matchers.`is`
import org.hamcrest.core.IsNot.not
import org.junit.*
import org.junit.runner.RunWith
import ru.iteco.fmhandroid.R
import ru.iteco.fmhandroid.page.LoginPage
import ru.iteco.fmhandroid.page.MainPage
import ru.iteco.fmhandroid.utils.Wait.forAnyDisplayed

@LargeTest
@RunWith(AndroidJUnit4::class)
class AuthTest {

    @Rule
    @JvmField
    var activityRule = ActivityScenarioRule(AppActivity::class.java)

    private lateinit var decorView: View

    @Before
    fun ensureLoggedOut() {
        activityRule.scenario.onActivity { decorView = it.window.decorView }

        forAnyDisplayed(
            withHint("Login"),
            withId(R.id.authorization_image_button),
            timeoutMs = ru.iteco.fmhandroid.utils.Wait.TIMEOUT_LONG
        )
        try {
            onView(withHint("Login")).check(matches(isDisplayed()))
            return
        } catch (_: NoMatchingViewException) {
        }

        MainPage.logout()

        forAnyDisplayed(withHint("Login"), timeoutMs = ru.iteco.fmhandroid.utils.Wait.TIMEOUT_LONG)
    }

    @Test
    fun tc001_successfulLoginShowsMainScreen() {
        forAnyDisplayed(withHint("Login"))
        LoginPage.assertOnScreen()
        LoginPage.typeLogin("login2")
        LoginPage.typePassword("password2")
        LoginPage.tapSignIn()

        MainPage.assertOpened()
    }

    @Test
    fun tc002_loginWithInvalidCredentialsShowsToast() {
        LoginPage.assertOnScreen()

        LoginPage.typeLogin("login")
        LoginPage.typePassword("pass")
        LoginPage.tapSignIn()

        var currentDecorView: View? = null
        activityRule.scenario.onActivity {
            currentDecorView = it.window.decorView
        }

        onView(withText("Something went wrong. Try again later."))
            .inRoot(withDecorView(not(`is`(currentDecorView))))
            .check(matches(isDisplayed()))
    }


    @Test
    fun tc003_signInRefusedShowsToastWhenFieldsEmpty() {
        LoginPage.assertOnScreen()
        onView(withHint("Login")).perform(replaceText(""))
        onView(withHint("Password")).perform(replaceText(""))
        LoginPage.tapSignIn()

        var currentDecorView: View? = null
        activityRule.scenario.onActivity {
            currentDecorView = it.window.decorView
        }

        onView(withText("Login and password cannot be empty"))
            .inRoot(withDecorView(not(`is`(currentDecorView))))
            .check(matches(isDisplayed()))
    }

    @Test
    fun tc012_logout_returnsToLogin() {
        LoginPage.assertOnScreen()
        LoginPage.typeLogin("login2")
        LoginPage.typePassword("password2")
        LoginPage.tapSignIn()
        MainPage.assertOpened()

        MainPage.logout()

        onView(withHint("Login")).check(matches(isDisplayed()))
    }

    @Test
    fun tc017_reloginAfterLogout_works() {
        LoginPage.assertOnScreen()
        LoginPage.typeLogin("login2")
        LoginPage.typePassword("password2")
        LoginPage.tapSignIn()
        MainPage.assertOpened()

        MainPage.logout()
        onView(withHint("Login")).check(matches(isDisplayed()))

        LoginPage.typeLogin("login2")
        LoginPage.typePassword("password2")
        LoginPage.tapSignIn()
        MainPage.assertOpened()
    }
}